File: 백준/Silver/11279. 최대 힙/최대 힙.cc
#include <iostream>
#include <cmath>

using namespace std;


int heap[200000];
int heapSize;

void upHeap(int q) {
	heap[1 + heapSize++] = q;
	int now = heapSize;
	while (now > 1 && heap[now / 2] < heap[now]) {
		swap(heap[now / 2], heap[now]);
		now /= 2;
	}
}

void downHeap() {
	if (!heapSize) {
		cout << 0 << '\n';
		return;
	}
	cout << heap[1] << '\n';
	heap[1] = heap[heapSize];
	heap[heapSize--] = 0;
	int now = 1;
	while (!(heap[now] >= heap[2 * now] && heap[now] >= heap[2 * now + 1])) {
		if (heap[2 * now] + heap[2 * now + 1] == 0) break;
		if (heap[2 * now] > heap[2 * now + 1]) {
			swap(heap[now], heap[2 * now]);
			now *= 2;
		}
		else {
			swap(heap[now], heap[2 * now + 1]);
			now = 2 * now + 1;
		}
	}
}

void showHeap() {
	cout << '\n';
	for (int i = 0; i < 5; i++) {
		for (int j = 0; j < (1 << i); j++) {
			int idx = pow(2, i) + j;
			cout << heap[idx] << ' ';
		}
		cout << '\n';
	}
}

int main(void) {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	heap[0] = -100;

	int n, q;
	cin >> n;
	for (int i = 0; i < n; i++) {
		cin >> q;
		if (q)
			upHeap(q);
		else
			downHeap();
		//showHeap();
	}


	return 0;
}

File: 백준/Silver/11279. 최대 힙/최대 힙.py
import sys
i=sys.stdin.readline

class MaxHeap:
    def __init__(self):
        self.h=[]
        self.size=0
    
    def upheap(self, n):
        self.h.append(n)
        self.size+=1
        now=self.size-1
        while now and self.h[(now-1)//2]<self.h[now]:
            self.h[(now-1)//2],self.h[now]=self.h[now],self.h[(now-1)//2]
            now=(now-1)//2
    
    def downheap(self):
        if self.size==0:print(0);return
        elif self.size==1:print(self.h.pop());self.size=0;return
        print(self.h[0])
        self.h[0]=self.h.pop()
        self.size-=1
        now=0
        while (2*now+1<self.size and self.h[now]<self.h[2*now+1]) \
        or (2*now+2<self.size and self.h[now]<self.h[2*now+2]):
            if 2*now+2>=self.size:
                self.h[now],self.h[2*now+1]=self.h[2*now+1],self.h[now]
                return
            if self.h[2*now+2]>self.h[2*now+1]:
                self.h[now],self.h[2*now+2]=self.h[2*now+2],self.h[now]
                now=2*now+2
            else:
                self.h[now],self.h[2*now+1]=self.h[2*now+1],self.h[now]
                now=2*now+1
    
    def showheap(self):
        print()
        #print(self.h)

        i=1
        for x in range(self.size):
            print(self.h[x],end=' ')
            if x+1==(1<<i)-1:print();i+=1
        print('\n')

heap=MaxHeap()
for x in '0'*int(i()):
    q=int(i())
    if q==0:heap.downheap()
    else:heap.upheap(q)
    #heap.showheap()